crates/proto/src/rr/rdata/tsig.rs

```diff
@@ -0,0 +1,770 @@
+///      +--------------------------+----------------+-----------------+
+///      | hmac-sha256              | MUST           | RECOMMENDED     |
+///      +--------------------------+----------------+-----------------+
+///      | hmac-sha256-128          | MAY            | MAY             |
+///      +--------------------------+----------------+-----------------+
+///      | hmac-sha384              | MAY            | MAY             |
+///      +--------------------------+----------------+-----------------+
+///      | hmac-sha384-192          | MAY            | MAY             |
+///      +--------------------------+----------------+-----------------+
+///      | hmac-sha512              | MAY            | MAY             |
+///      +--------------------------+----------------+-----------------+
+///      | hmac-sha512-256          | MAY            | MAY             |
+///      +--------------------------+----------------+-----------------+
+/// ```
+#[derive(Debug, PartialEq, Eq, Hash, Clone)]
+pub enum Algorithm {
+    /// HMAC-MD5.SIG-ALG.REG.INT (not supported for cryptographic operations)
+    HmacMd5,
+    /// gss-tsig (not supported for cryptographic operations)
+    Gss,
+    /// hmac-sha1 (not supported for cryptographic operations)
+    HmacSha1,
+    /// hmac-sha224 (not supported for cryptographic operations)
+    HmacSha224,
+    /// hmac-sha256
+    HmacSha256,
+    /// hmac-sha256-128 (not supported for cryptographic operations)
+    HmacSha256_128,
+    /// hmac-sha384
+    HmacSha384,
+    /// hmac-sha384-192 (not supported for cryptographic operations)
+    HmacSha384_192,
+    /// hmac-sha512
+    HmacSha512,
+    /// hmac-sha512-256
+    HmacSha512_256,
+    /// Unkown algorithm
+    Unknown(Name),
+}
+
+impl TSIG {
+    /// Constructs a new TSIG
+    ///
+    /// [RFC 8945, Secret Key Transaction Authentication for DNS](https://tools.ietf.org/html/rfc8945#section-4.1)
+    ///
+    /// ```text
+    /// 4.1.  TSIG RR Type
+    ///
+    ///   To provide secret key authentication, we use an RR type whose
+    ///   mnemonic is TSIG and whose type code is 250.  TSIG is a meta-RR and
+    ///   MUST NOT be cached.  TSIG RRs are used for authentication between DNS
+    ///   entities that have established a shared secret key.  TSIG RRs are
+    ///   dynamically computed to cover a particular DNS transaction and are
+    ///   not DNS RRs in the usual sense.
+    ///
+    ///   As the TSIG RRs are related to one DNS request/response, there is no
+    ///   value in storing or retransmitting them; thus, the TSIG RR is
+    ///   discarded once it has been used to authenticate a DNS message.
+    /// ```
+    pub fn new(
+        algorithm: Algorithm,
+        time: u64,
+        fudge: u16,
+        mac: Vec<u8>,
+        oid: u16,
+        error: u16,
+        other: Vec<u8>,
+    ) -> Self {
+        // Should maybe not be a panic but return a Result::Err, or be ignored?
+        assert!(time < (1 << 48));
```

Comment:
Is `new` used during deserialization of data off the wire? If so, I think we should instead make this fallible and return a `Result`, otherwise this could be used to cause a DOS on clients (and potentially servers in the future).

Reply:
this is not used during deserialization iirc, and it would not matter as the field in the dns wire protocol is actually 48 bits (read as 16u + 32u, and "concatenated" to make a u64 in the code I wrote).
Same goes for `mac.len()` and `other.len()`, the wire protocol represent them as a u16 (len of buffer), followed by that much bytes of mac/other data.
They could be made into errors, but these assertions can only be triggered by library missuses, definitely not invalid packets

Reply:
Yeah at a minimum, can you add a Panic section to the API doc?

Reply:
I ended up removing the assertions as using `try_into` in emit made them redundant